aspnetcore/fundamentals/url-rewriting.md

```diff
@@ -0,0 +1,295 @@
+
+[!code-csharp[Main](url-rewriting/sample/Startup.cs?name=snippet1&highlight=3)]
+
+Original Request: **http://localhost:5000/rewrite-rule/1234/5678**
+
+![Browser window with Developer Tools tracking the request and response](url-rewriting/_static/add_rewrite.png)
+
+The first thing you will notice in the regex is the carat (`^`) at the beginning of the expression. This means that matching should be attempted starting at the beginning of the URL path.
+
+In the earlier example with the redirect rule, `redirect-rule/(.*)`, there is no carat at the start of the regex; therefore, any characters may precede **redirect-rule/** in the path for a successful match.
+
+URL | Match
+--- | :---:
+**http://localhost:5000/redirect-rule/1234/5678** | Yes
+**http://localhost:5000/my-cool-redirect-rule/1234/5678** | Yes
+**http://localhost:5000/foobarredirect-rule/1234/5678** | Yes
+
+The rewrite rule, `^rewrite-rule/(\d+)/(\d+)`, will only match paths if they start with **rewrite-rule/**. Notice the difference in matching between the rewrite rule below and the redirect rule above.
+
+URL | Match
+--- | :---:
+**http://localhost:5000/rewrite-rule/1234/5678** | Yes
+**http://localhost:5000/my-cool-rewrite-rule/1234/5678** | No
+**http://localhost:5000/foobarrewrite-rule/1234/5678** | No
+
+Following the `^rewrite-rule/` portion of the expression, there are two capture groups, `(\d+)/(\d+)`. The `\d` signifies *match a digit (number)*. The plus sign (`+`) means *match one or more of the preceding character*. Therefore, the URL must contain a number followed by a forward-slash followed by another number. These capture groups are injected into the resultant rewritten URL as `$1` and `$2`. The rewrite rule replacement string places the captured groups into the querystring. The requested URL of **http://localhost:5000/rewrite-rule/1234/5678** is rewritten to obtain the resource at **http://localhost:5000/rewritten?var1=1234&var2=5678**. If a querystring is present on the original request, it's preserved when the URL is rewritten.
+
+There is no roundtrip to the server to obtain the resource. If the resource exists, it's fetched and returned to the client with a 200 (OK) status code. Because the client isn't redirected, the URL in the browser address bar doesn't change. As far as the client is concerned, the URL rewrite operation never occurred.
+
+> [!NOTE]
+> `skipRemainingRules: true` should be used whenever possible, because matching rules is an expensive process and slows down application response time. For the fastest application response, order your rewrite rules from most frequently matched to least frequently matched and skip the processing of the remaining rules when a match occurs and no additional rule processing is required.
+
+#### Apache mod_rewrite
```

Comment:
I don't make any assumption at this point about why the dev might not want to use the module on Apache or IIS. We say above in L54 that we recommend that the server modules be used, due to feature support and perf. Do you think L54 isn't enough?